decode.go

```diff
@@ -123,166 +135,234 @@ func parseDateTime(b []byte) (time.Time, error) {
 	}
+
 	dt.Date = date
 
-	sep := b[10]
+	sep := b[localDateTimeByteLen-1]
 	if sep != 'T' && sep != ' ' {
-		return dt, nil, fmt.Errorf("datetime separator is expected to be T or a space")
+		return dt, nil, ErrParseLocalDateTimeWrongSeparator
 	}
 
 	t, rest, err := parseLocalTime(b[11:])
 	if err != nil {
 		return dt, nil, err
 	}
+
 	dt.Time = t
 
 	return dt, rest, nil
 }
 
+var ErrParseLocalTimeWrongLength = errors.New("times are expected to have the format HH:MM:SS[.NNNNNN]")
+
+const localTimeByteLen = 8
+
 // parseLocalTime is a bit different because it also returns the remaining
 // []byte that is didn't need. This is to allow parseDateTime to parse those
 // remaining bytes as a timezone.
 func parseLocalTime(b []byte) (LocalTime, []byte, error) {
-	t := LocalTime{}
+	var t LocalTime
 
-	if len(b) < 8 {
-		return t, nil, fmt.Errorf("times are expected to have the format HH:MM:SS[.NNNNNN]")
+	if len(b) < localTimeByteLen {
+		return t, nil, ErrParseLocalTimeWrongLength
 	}
 
 	var err error
+
 	t.Hour, err = parseDecimalDigits(b[0:2])
 	if err != nil {
 		return t, nil, err
 	}
+
 	if b[2] != ':' {
 		return t, nil, newDecodeError(b[2:3], "expecting colon between hours and minutes")
 	}
+
 	t.Minute, err = parseDecimalDigits(b[3:5])
 	if err != nil {
 		return t, nil, err
 	}
+
 	if b[5] != ':' {
 		return t, nil, newDecodeError(b[5:6], "expecting colon between minutes and seconds")
 	}
+
 	t.Second, err = parseDecimalDigits(b[6:8])
 	if err != nil {
 		return t, nil, err
 	}
 
 	if len(b) >= 15 && b[8] == '.' {
 		t.Nanosecond, err = parseDecimalDigits(b[9:15])
+		if err != nil {
+			return t, nil, err
+		}
+
 		return t, b[15:], nil
 	}
 
 	return t, b[8:], nil
 }
 
+var (
+	ErrParseFloatStartDot = errors.New("float cannot start with a dot")
+	ErrParseFloatEndDot   = errors.New("float cannot end with a dot")
+)
+
+//nolint:cyclop
 func parseFloat(b []byte) (float64, error) {
+	//nolint:godox
 	// TODO: inefficient
 	if len(b) == 4 && (b[0] == '+' || b[0] == '-') && b[1] == 'n' && b[2] == 'a' && b[3] == 'n' {
 		return math.NaN(), nil
 	}
 
 	tok := string(b)
-	err := numberContainsInvalidUnderscore(tok)
-	if err != nil {
+	if err := numberContainsInvalidUnderscore(tok); err != nil {
 		return 0, err
 	}
+
 	cleanedVal := cleanupNumberToken(tok)
 	if cleanedVal[0] == '.' {
-		return 0, fmt.Errorf("float cannot start with a dot")
+		return 0, ErrParseFloatStartDot
 	}
+
 	if cleanedVal[len(cleanedVal)-1] == '.' {
-		return 0, fmt.Errorf("float cannot end with a dot")
+		return 0, ErrParseFloatEndDot
+	}
+
+	f, err := strconv.ParseFloat(cleanedVal, 64)
+	if err != nil {
+		return 0, fmt.Errorf("coudn't ParseFloat %w", err)
 	}
-	return strconv.ParseFloat(cleanedVal, 64)
+
+	return f, nil
 }
 
 func parseIntHex(b []byte) (int64, error) {
 	tok := string(b)
 	cleanedVal := cleanupNumberToken(tok)
-	err := hexNumberContainsInvalidUnderscore(cleanedVal)
+
+	if err := hexNumberContainsInvalidUnderscore(cleanedVal); err != nil {
```

Comment:
I would like to keep the function call and error handling separate. I find the (arguable) readability and the consistency with functions returning more than one value to outweigh the conciseness and scope of err.

Reply:
actually, I re-ran the linter, and it advocates for this on only specific cases. So it seems I have been a bit overzealous on this one.
I will rerun it as well.